moveit_ros/hybrid_planning/local_planner/local_constraint_solver_plugins/src/servo_solver.cpp

```diff
@@ -0,0 +1,174 @@
+
+  // Use for debugging
+  // twist_cmd_pub_ = node_->create_publisher<geometry_msgs::msg::TwistStamped>("~/delta_twist_cmds", 10);
+  return true;
+}
+
+bool ServoSolver::reset()
+{
+  RCLCPP_INFO(getLogger(), "Reset Servo Solver");
+  joint_cmd_rolling_window_.clear();
+  return true;
+};
+
+moveit_msgs::action::LocalPlanner::Feedback
+ServoSolver::solve(const robot_trajectory::RobotTrajectory& local_trajectory,
+                   const std::shared_ptr<const moveit_msgs::action::LocalPlanner::Goal> /*local_goal*/,
+                   trajectory_msgs::msg::JointTrajectory& local_solution)
+{
+  // Create Feedback
+  moveit_msgs::action::LocalPlanner::Feedback feedback_result;
+
+  // Transform next robot trajectory waypoint into JointJog message
+  moveit_msgs::msg::RobotTrajectory robot_command;
+  local_trajectory.getRobotTrajectoryMsg(robot_command);
+
+  if (robot_command.joint_trajectory.points.empty())
+  {
+    feedback_result.feedback = std::string("Reference trajectory does not contain any points");
+    return feedback_result;
+  }
+
+  // Get current state
+  const auto current_state = planning_scene_monitor_->getStateMonitor()->getCurrentState();
+
+  // Create goal state
+  moveit::core::RobotState target_state(local_trajectory.getRobotModel());
+  target_state.setVariablePositions(robot_command.joint_trajectory.joint_names,
+                                    robot_command.joint_trajectory.points[0].positions);
+  target_state.update();
+
+  // TF planning_frame -> current EE
+  Eigen::Isometry3d current_pose = current_state->getFrameTransform(solver_parameters_.reference_frame);
+  // TF planning -> target EE
+  Eigen::Isometry3d target_pose = target_state.getFrameTransform(solver_parameters_.reference_frame);
+
+  // current EE -> planning frame * planning frame -> target EE
+  Eigen::Isometry3d diff_pose = current_pose.inverse() * target_pose;
+  Eigen::AngleAxisd axis_angle(diff_pose.linear());
+
+  constexpr double fixed_trans_vel = 0.05;
+  constexpr double fixed_rot_vel = 5;
+  const double trans_gain = fixed_trans_vel / diff_pose.translation().norm();
+  const double rot_gain = fixed_rot_vel / diff_pose.rotation().norm();
+
+  // Calculate Cartesian command delta
+  // Transform current pose to command frame
+  // Transform goal pose to command frame
+  servo_->setCommandType(moveit_servo::CommandType::TWIST);
+  moveit_servo::TwistCommand target_twist{
+    solver_parameters_.reference_frame,
+    { diff_pose.translation().x() * trans_gain, diff_pose.translation().y() * trans_gain,
+      diff_pose.translation().z() * trans_gain, axis_angle.axis().x() * axis_angle.angle() * rot_gain,
+      axis_angle.axis().y() * axis_angle.angle() * rot_gain, axis_angle.axis().z() * axis_angle.angle() * rot_gain }
+  };
+
+  std::optional<trajectory_msgs::msg::JointTrajectory> trajectory_msg;
+  // Create servo commands until a trajectory message can be generated
+  while (!trajectory_msg)
```

Comment:
Before beginning the loop, delete any values in the command queue that are in than 1 update period in the past. So if the local planner is called at 100 Hz, then keep values as long as they are not older than 0.01 seconds.